include/fiction/technology/sidb_charge_state.hpp

```diff
@@ -122,6 +123,123 @@ charge_configuration_to_string(const std::vector<sidb_charge_state>& charge_dist
 
     return config_str.str();
 }
+/**
+ * The iteration direction over the SiDB states.
+ */
+enum class sidb_state_iter_dir
+{
+    /**
+     * To positive iterates "\f$-\to 0\to +\f$".
+     */
+    TO_POSITIVE,
+    /**
+     * To negative iterates "\f$+\to 0\to -\f$".
+     */
+    TO_NEGATIVE
+};
+/**
+ * Iterator for SiDB charge states. Iterates `-1 -> 0 -> +1 -> NONE`.
+ */
+template <sidb_state_iter_dir dir = sidb_state_iter_dir::TO_POSITIVE>
+class sidb_charge_state_iterator
+{
+  public:
+    // Iterator typedefs
+    using iterator_category = std::input_iterator_tag;
```

Comment:
What are they for?

Reply:
So you can use the iterator like a C++ iterator like shown in the comment above (range-based)